Remove auto-discovery from specific DB guides

ptgott opened the pull request:
Closes #22942

Help users enroll specific databases with Teleport while directing them auto-discovery instructions if they want to go that route.

- Revert #30913 for guides to enrolling specific databases. For guides that always included auto-discovery instructions, add instructions to enroll databases at specific URLs.
- Add a note in the introduction sections of AWS guides that it is possible to enroll AWS databases with auto-discovery, linking to the relevant guide.

ptgott committed:
Remove auto-discovery from specific DB guides

Closes #22942

Help users enroll specific databases with Teleport while directing them
auto-discovery instructions if they want to go that route.

- Revert #30913 for guides to enrolling specific databases. For guides
  that always included auto-discovery instructions, add instructions to
  enroll databases at specific URLs.
- Add a note in the introduction sections of AWS guides that it is
  possible to enroll AWS databases with auto-discovery, linking to the
  relevant guide.

ptgott commented:
Note for reviewers: this PR is based on top of https://github.com/gravitational/teleport/pull/38574.

greedy52 commented:
Thanks @ptgott! I was also discussing with the team the other day that it would be better to do static urls for starter guides and then direct them to auto-discovery separately. I will take a closer look of this PR next week.

greedy52 commented on docs/pages/database-access/enroll-aws-databases/redshift-serverless.mdx:
```diff
@@ -0,0 +1,363 @@
+  EC2 instance, and provides a corresponding example of access control.
+- (!docs/pages/includes/tctl.mdx!)
+
+## Step 1/5. Create an IAM policy for Teleport
```
IIRC, this section was written by Alex to list out all the IAM permissions that are required for running the bootstrap command. 

I would suggest removing this step to be more aligned with other guides where we don't show this detail, assuming most users using the bootstrap command are admins. 

I can add this to the AWS IAM reference page when I work on it. WDYT?

ptgott replied:
It looks like the AWS database guides are pretty inconsistent here. Some guides include the `docs/pages/includes/database-access/aws-bootstrap.mdx` partial, which includes both manual and automatic options for the `teleport db configure bootstrap` command, but don't include an IAM policy to enable automatic IAM bootstrapping.

Would it make sense to add the IAM policy for automatic bootstrapping to the `aws-bootstrap.mdx` partial? If it makes more sense, we could also add the policy to another partial and include that partial in all of the pages that use the `aws-bootstrap.mdx` partial. We could then edit this page to follow the same structure.

How does that sound?

greedy52 replied:
> Would it make sense to add the IAM policy for automatic bootstrapping to the aws-bootstrap.mdx partial? If it makes more sense, we could also add the policy to another partial and include that partial in all of the pages that use the aws-bootstrap.mdx partial. We could then edit this page to follow the same structure.
> How does that sound?

Sure. It would be beneficial for those who want to know the exact IAM permissions that are required for bootstrapping. But I wouldn't create a new IAM role/user with those permissions just to run the bootstrap command. Normally you should just run bootstrapping as an admin. 

> It looks like the AWS database guides are pretty inconsistent here.

Agreed. There is also a whole separate discussion about whether we should suggest bootstrapping at all since it's not very easy to use. Bootstrapping may be ok for a simple static URL like the new guide. But bootstrapping is very complicated for auto-discovery and cross-account support. I was planning to address that separately, either always using `teleport db configure aws print-iam`  or just listing out the exact permissions in IAM JSON format.

ptgott replied:
That makes sense! Just to make sure I understand, the recommended move across the AWS database access guides is to:
- Only show manual bootstrap command instructions
- Require that the user have admin permissions on AWS
- Remove IAM policies for bootstrapping

greedy52 replied:
> Remove IAM policies for bootstrapping
Yes

> Only show manual bootstrap command instructions
> Require that the user have admin permissions on AWS

These two are either-or. If we require the reader to have admin permissions, then they can just use "automatic" bootstrap. 

For the next round, I guess i will list out the actual IAM permissions required for each database type in each guide similar to this https://goteleport.com/docs/auto-discovery/servers/ec2-discovery/#step-27-define-an-iam-policy, and have a separate guide/reference on bootstrapping

ptgott replied:
Sounds good! 👍

greedy52 commented on docs/pages/database-access/enroll-aws-databases/redshift-serverless.mdx:
```diff
@@ -0,0 +1,363 @@
+    protocol: postgres
+    aws:
+      redshift_serverless:
+        workgroup_name: WORKGROUP_NAME
+        endpoint_name: ENDPOINT_NAME
```
`WORKGROUP_NAME` and `ENDPOINT_NAME` are both optional (as they can be parsed from URI as well). `ENDPOINT_NAME` is only for VPC endpoints, so regular workgroups don't need it.

Alternatively, we can also use `teleport db configure create` to generate the db config, with protocol and URI.

greedy52 commented on docs/pages/database-access/enroll-aws-databases/redshift-serverless.mdx:
````diff
@@ -0,0 +1,363 @@
+```code
+$ tsh login --proxy=mytenant.teleport.sh --user=alice
+$ tsh db ls
+Name        Description                    Labels
+----------- ------------------------------ --------
+my-redshift Redshift cluster in us-east-1  ...
````
The description here could be a problem for all updated AWS guides. `xxxx in us-east-1` description is added by auto-discovery. With `teleport db configure create`, the description should be empty.

greedy52 commented on docs/pages/database-access/enroll-aws-databases/sql-server-ad.mdx:
```diff
@@ -18,6 +18,11 @@ videoBanner: k2wz79XCexY
 This guide will focus on Amazon RDS for SQL Server using AWS-managed Active
 Directory authentication.
 
+This guide shows how to enroll a single Microsoft SQL Server database through
+the Teleport Database Service. For a more scalable approach, learn how to
+[configure the Teleport Discovery Service](../../auto-discovery/databases.mdx)
+to automatically enroll all AWS databases in your infrastructure.
```
```suggestion

```

There is no auto-discovery for RDS SQL server with AD.

(not to be confused with RDS Proxy for SQL Server).

greedy52 commented on docs/pages/database-access/enroll-azure-databases.mdx:
```diff
@@ -0,0 +1,11 @@
+You can protect Azure-managed databases with Teleport. Learn how to enroll the
+following databases:
+
+- [Azure SQl Server](./enroll-azure-databases/azure-sql-server-ad.mdx)
```
```suggestion
- [Azure SQL Server](./enroll-azure-databases/azure-sql-server-ad.mdx)
```

greedy52 commented on docs/pages/database-access/enroll-azure-databases.mdx:
```diff
@@ -0,0 +1,11 @@
+following databases:
+
+- [Azure SQl Server](./enroll-azure-databases/azure-sql-server-ad.mdx)
+- [Azure Database for PostgreSQL or MySQL Server](./enroll-azure-databases/azure-postgres-mysql.mdx)
```
We probably can break it down into two separate guides in the future =D.

greedy52 commented on docs/pages/database-access/introduction.mdx:
```diff
@@ -39,14 +39,38 @@ with GitHub, execute a few SQL queries and observe them in the audit log:
+Teleport can protect both self-hosted and managed databases. Learn how to set up
+secure access to databases in your infrastructure:
+
+- Managed [Amazon Web Services](./enroll-aws-databases.mdx) databases
+- Managed [Google Cloud](./enroll-google-cloud-databases.mdx) databases
+- [Self-Hosted](./enroll-self-hosted-databases.mdx) databases, regardless of
+  cloud provider
+- Managed [Microsoft Azure](./enroll-azure-databases.mdx) databases
+- [Database-specific cloud
+  platforms](docs/pages/database-access//enroll-managed-databases.mdx) such as
+  Snowflake and MongoDB Atlas
```
I only see cloud-hosted in this list. do we want to add self-hosted link?

ptgott replied:
I addressed this in #38574 and incorporated it when rebasing

greedy52 commented on docs/pages/includes/config-reference/database-config.yaml:
```diff
@@ -24,7 +24,6 @@ db_service:
     # 'redshift-serverless' - discovers and registers AWS Redshift Serverless databases.
     # 'elasticache' - discovers and registers AWS ElastiCache Redis databases.
     # 'memorydb' - discovers and registers AWS MemoryDB Redis databases.
-    # 'opensearch' - discovers and registers AWS OpenSearch Redis databases.
```
`opensearch` does exist. https://github.com/gravitational/teleport/blob/73757d829da99c4f2d9773c221d7a5484d4ba539/api/types/matchers_aws.go#L61-L62

I guess we should add it in the next line. Thanks for catching this!

greedy52 commented on docs/pages/includes/database-access/aws-bootstrap.mdx:
```diff
@@ -1,35 +1,5 @@
-
-<Tabs>
-
-<TabItem label="Auto-discovery">
```
👍

greedy52 commented on docs/pages/includes/database-access/rds-proxy.mdx:
```diff
@@ -32,16 +36,44 @@ Teleport currently supports RDS Proxy instances with engine family
+   -o file \
+   --proxy=<Var name="teleport.example.com"/>:443 \
+   --token=/tmp/token \
+   --rdsproxy-discovery=<Var name="us-west-1" />
```
same as above, `--url` + `--protocol` would do.

ptgott committed:
Respond to greedy52 feedback

- Remove auto-discovery notes where they are incorrect.
- Remove unneeded `id` flags.
- Use `teleport db configure create` for Redshift Serverless, removing
  unneeded config fields. Also add the `aws-bootstrap.mdx` partial to
  the Redshift Serverless guide, making it more consistent with other
  guides that include `teleport db configure bootstrap` instructions.
- Minor spelling/grammar tweaks.
- Fix example description outputs.
- Use URI/protocol flags in `rds-proxy.mdx` instructions.
- Restore `opensearch` to `database-config.yaml`.

ptgott commented:
I've responded to some feedback in d5c971ce9d61413cb5c2d216aacda1752a40b1f5. I'll continue to push commits until I've finished!

greedy52 commented on docs/pages/database-access/enroll-aws-databases/rds.mdx:
````diff
@@ -136,7 +140,50 @@ Next, get your environment ready to run the Teleport Database Service:
+```code
+$ sudo teleport db configure create \
+   -o file \
+   --proxy=<Var name="example.teleport.sh:443" />  \
````
Sorry if i missed this earlier, `teleport db configure create` requires a `--name` for the database name to be registered like `postgres-rds`.

Nit: One optional parameter we could also add is `--labels env=dev`. It's a good practice to have some labels for the database. However, since we currently assigning the default `access` role which matches wildcard labels, `--labels` is not necessary. If we do add the `--labels env=dev`, we will have to add like `env=dev` in the `Labels` column of `tsh db ls`. So probably easier without it.

greedy52 commented on docs/pages/database-access/enroll-aws-databases/rds.mdx:
```diff
@@ -388,21 +429,12 @@ $ tsh login --proxy=<Var name="example.teleport.sh:443" /> --user=alice
-# aurora-mysql                   Aurora cluster in us-west-1                   ...
-# aurora-mysql-custom-myendpoint Aurora cluster in us-west-1 (custom endpoint) ...
-# aurora-mysql-reader            Aurora cluster in us-west-1 (reader endpoint) ...
+# postgres-rds                                                                 ...
+# aurora-mysql                                                                 ...
+# aurora-mysql-custom-myendpoint                                               ...
+# aurora-mysql-reader                                                          ...
```
nit: wonder if we should just keep the single database that's been configured and get rid other endpoints from discovery. also `tsh db ls` will "shrink" the description column when there is nothing. 

```suggestion
$ tsh db ls
# Name         Description Labels
# ------------ ----------- --------
# postgres-rds 
```

greedy52 commented on docs/pages/database-access/enroll-aws-databases/redis-aws.mdx:
```diff
@@ -38,22 +43,50 @@ description: How to configure Teleport database access with AWS ElastiCache and
+     -o file \
+     --proxy=example.teleport.sh:443 \
+     --protocol="redis" \
+     --uri=${ELASTICACHE_URI?} \
```
```suggestion
     --uri=${ELASTICACHE_URI} \
```
?

ptgott replied:
This instructs the shell to raise an error if the variable isn't defined

greedy52 reviewed (approved):
overall LGTM. Left on comment only adding `--name` to `teleport db configure create`

ptgott committed:
Respond to greedy52 feedback

- Add the `--name` parameter to `db configure create`
- Edit `tsh db ls` output in the RDS guide

ptgott commented:
@greedy52 Since the original base branch, `paul.gottschling/17641-db-org`, has taken awhile to get reviewed, I've changed the base to `master`.

ptgott committed:
Fix linter issues
